File: typescript/src/mz.ts
import x86_16 from "./byte-operation/x86_16";
import ByteStream from "./bytestream";

/**
 * Named after Mark Zbikowski, MZ is the main MS-DOS EXE format found in all executables and libraries, containing basic information of the whole executable.
 */
export class MZ {
    /**
     * The Signature, or "Magic" number that always starts the MZ header.
     */
    Signature: "MZ";
    /**
     * Number of bytes in the last page.
     */
    LastPageBytes: number;
    /**
     * Number of whole/partial pages.
     */
    PageCount: number;
    /**
     * Number of entries in the relocation table
     */
    RelocationTableEntryCount: number;
    /**
     * The number of paragraphs taken up by the header. It can be any value, as the loader just uses it to find where the actual executable data starts.
     * 
     * It may be larger than what the "standard" fields take up, and you may use it if you want to include your own header metadata, or put the relocation table there, or use it for any other purpose.
     */
    HeaderSize: number;
    /**
     * The number of paragraphs **required** by the program, excluding the PSP and program image. If no free block is big enough, the loading stops.
     */
    MinimumAllocation: number;
    /**
     * The number of paragraphs **requested** by the program. If no free block is big enough, the biggest one possible is allocated.
     */
    MaximumAllocation: number;
    /**
     * Relocatable segment address for SS.
     */
    InitialSS: number;
    /**
     * Initial value for SP.
     */
    InitialSP: number;
    /**
     * When added to the sum of all other words in the file, the result should be zero.
     */
    Checksum: number;
    /**
     * Initial value for IP.
     */
    InitialIP: number;
    /**
     * Relocatable segment address for CS.
     */
    InitialCS: number;
    /**
     * The (absolute) offset to the relocation table.
     */
    RelocationTableOffset: number;
    /**
     * Value used for overlay management. If zero, this is the main executable.
     */
    Overlay: number;

    /**
     * The OEM Id.  This value is only read in Windows executables.
     */
    OEMIdentifier?: number;
    /**
     * The OEM Info.  This value is only read in Windows executables.
     */
    OEMInfo?: number;
    /**
     * The starting address of the new header.
     */
    NewHeaderStart?: number;
    /**
     * The relocation tables.
     */
    RelocationTables: RelocationTable[];

    /**
     * The assembly code.
     */
    HeaderCode: string[];

    private constructor(magic: "MZ",
        lastPageBytes: number,
        pageCount: number,
        numEntriesRelocTable: number,
        headerSize: number,
        minAlloc: number,
        maxAlloc: number,
        initSS: number,
        initSP: number,
        checksum: number,
        initIP: number,
        initCS: number,
        relocTableOffset: number,
        overlay: number,
        headerCode: string[],
        relocTbls: RelocationTable[],
        oemid?: number,
        oeminfo?: number,
        newHeaderStart?: number
    ) {
        this.Signature = magic;
        this.LastPageBytes = lastPageBytes;
        this.PageCount = pageCount;
        this.RelocationTableEntryCount = numEntriesRelocTable;
        this.HeaderSize = headerSize;
        this.MinimumAllocation = minAlloc;
        this.MaximumAllocation = maxAlloc;
        this.InitialSS = initSS;
        this.InitialSP = initSP;
        this.Checksum = checksum;
        this.InitialIP = initIP;
        this.InitialCS = initCS;
        this.RelocationTableOffset = relocTableOffset;
        this.Overlay = overlay;
        this.OEMIdentifier = oemid;
        this.OEMInfo = oeminfo;
        this.NewHeaderStart = newHeaderStart;
        this.HeaderCode = headerCode;
        this.RelocationTables = relocTbls;
    }

    /**
     * Reads a byte stream.  Throws an error is the first two bytes don't match "MZ"
     * @param bst The byte stream
     * @returns An instance of the MZ class
     */
    public static Read(bst: ByteStream) {
        if (bst.ReadString(2) != "MZ") {
            throw new Error();
        }

        const lastPageBytes = bst.ReadWord();
        const pageCount = bst.ReadWord();
        const numEntriesRelocTable = bst.ReadWord();
        const headerSize = bst.ReadWord();
        const minAlloc = bst.ReadWord();
        const maxAlloc = bst.ReadWord();
        const initSS = bst.ReadWord();
        const initSP = bst.ReadWord();
        const checksum = bst.ReadWord();
        const initIP = bst.ReadWord();
        const initCS = bst.ReadWord();
        const relocTableOffset = bst.ReadWord();
        const overlay = bst.ReadWord();

        let oemid: number | undefined = undefined;
        let oeminfo: number | undefined = undefined;
        let newHeaderStart: number | undefined = undefined;
        bst.Position += 8; // skip reserved
        oemid = bst.ReadWord();
        oeminfo = bst.ReadWord();
        bst.Position += 20; // skip reserved
        newHeaderStart = bst.ReadDWord();

        let relocTbls: RelocationTable[] = [];
        if (bst.Position == relocTableOffset && numEntriesRelocTable > 0) {
            for (let i = 0; i < numEntriesRelocTable; i++) {
                relocTbls.push(RelocationTable.Read(bst));
            }
        }

        if (bst.Position < headerSize * 16) {
            if (!bst.CheckReserved((headerSize * 16) - bst.Position)) {
                debugger;
            }
        }

        if (!newHeaderStart) {
            throw new Error();
        }

        x86_16.ss = initSS;
        x86_16.sp = initSP;
        x86_16.ip = initIP;
        x86_16.cs = initCS;
        let bytes = bst.ReadBytes(newHeaderStart - headerSize * 16);
        let code = x86_16.parseCode(bytes);

        return new MZ("MZ", lastPageBytes, pageCount, numEntriesRelocTable, headerSize, minAlloc, maxAlloc, initSS, initSP, checksum, initIP, initCS, relocTableOffset, overlay, code, relocTbls, oemid, oeminfo, newHeaderStart);
    }
}

export class RelocationTable {
    Offset: number;
    Segment: number;

    private constructor(offset: number, segment: number) {
        this.Offset = offset;
        this.Segment = segment;
    }

    public static Read(bst: ByteStream) {
        return new RelocationTable(bst.ReadWord(), bst.ReadWord());
    }
}

File: rust/src/mz.rs
use crate::{byte_stream::ByteStream, executable::Signature};

pub struct MZ {
    pub last_page_bytes: u16,
    pub page_count: u16,
    pub relocation_table_entry_count: u16,
    pub header_size: u16,
    pub min_alloc: u16,
    pub max_alloc: u16,
    pub init_ss: u16,
    pub init_sp: u16,
    pub checksum: u16,
    pub init_ip: u16,
    pub init_cs: u16,
    pub relocation_table_offset: u16,
    pub overlay: u16,

    pub oem_id: Option<u16>,
    pub oem_info: Option<u16>,
    pub new_header_start: Option<u32>,

    pub relocation_tables: Vec<RelocationTable>,
    pub header_code: Vec<String>,
}

impl MZ {
    pub fn signature() -> Signature {
        Signature::MZ
    }

    pub fn read(bst: &mut ByteStream) -> Self {
        let last_page_bytes = bst.read_word();
        let page_count = bst.read_word();
        let relocation_table_entry_count = bst.read_word();
        let header_size = bst.read_word();
        let min_alloc = bst.read_word();
        let max_alloc = bst.read_word();
        let init_ss = bst.read_word();
        let init_sp = bst.read_word();
        let checksum = bst.read_word();
        let init_ip = bst.read_word();
        let init_cs = bst.read_word();
        let relocation_table_offset = bst.read_word();
        let overlay = bst.read_word();

        let mut oem_id = None;
        let mut oem_info = None;
        let mut new_header_start = None;
        bst.pos += 8;//bst.check_reserved(8); // skip instead of throw, for linker compatibility reasons
        oem_id = Some(bst.read_word());
        oem_info = Some(bst.read_word());
        bst.pos += 8;//bst.check_reserved(20);
        new_header_start = Some(bst.read_dword());

        let mut relocation_tables = Vec::new();
        if bst.pos == relocation_table_offset as usize && relocation_table_entry_count > 0 {
            for _ in 0..relocation_table_entry_count {
                relocation_tables.push(RelocationTable::read(bst));
            }
        }

        if bst.pos < header_size as usize * 16 {
            if !bst.check_reserved((header_size as usize * 16) - bst.pos) {
                panic!();
            }
        }

        new_header_start.unwrap(); // more like a "todo!();" call

        

        MZ {
            last_page_bytes,
            page_count,
            relocation_table_entry_count,
            header_size,
            min_alloc,
            max_alloc,
            init_ss,
            init_sp,
            checksum,
            init_ip,
            init_cs,
            relocation_table_offset,
            overlay,

            oem_id,
            oem_info,
            new_header_start,
            relocation_tables,
            header_code: Vec::new(),
        }
    }
}

pub struct RelocationTable {
    pub offset: u16,
    pub segment: u16,
}

impl RelocationTable {
    pub fn read(bst: &mut ByteStream) -> Self {
        let offset = bst.read_word();
        let segment = bst.read_word();
        Self { offset, segment }
    }
}
